src/runtime/module_init_wrapper.cc

```diff
@@ -0,0 +1,234 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/runtime/module_init_wrapper.cc
+ * \brief A wrapper for initializing modules using metadata
+ */
+#include <tvm/node/container.h>
+#include <tvm/runtime/ndarray.h>
+#include <tvm/runtime/packed_func.h>
+#include <tvm/runtime/registry.h>
+
+#include <cstdint>
+#include <sstream>
+
+#include "file_util.h"
+
+namespace tvm {
+namespace runtime {
+
+using StringNDArrayMap = std::unordered_map<String, runtime::NDArray, ObjectHash, ObjectEqual>;
+
+class CSourceMetadataInitializer {
+ public:
+  explicit CSourceMetadataInitializer(const StringNDArrayMap& metadata) : metadata_(metadata) {}
+
+  template <typename T>
```

Comment:
Do we really need this one? My guess is that we can return MetaInitWrapper(that wraps CSourceModule) and our module serialization mechanism should take care of loading the meta-data back, without having to serialie the meta data to the C source.

Reply:
see also https://tvm.apache.org/docs/dev/introduction_to_module_serialization.html

Reply:
This is needed for partitioned subgraphs because they expect the plain c source data as the input.

Reply:
Can we change the subgraph generator to instead expose `InitMod0Func0(Array<NDArray>)` instead? so that we don't have to compile the meta data into the C source?

Reply:
This should be possible. Let me spend some time to refactor the c source code generator over the weekend.